allennlp/tests/interpret/hotflip_test.py

```diff
@@ -30,53 +30,6 @@ def test_hotflip(self):
             attack["original"]
         )  # hotflip replaces words without removing
 
-        # test using SQuAD model (tests different equals method)
```

Comment:
Can you add this to a test somewhere in the RC repo?  This is an argument for pulling the interpret code into its own repo, so we can actually do integration tests like this, but there are things on the `Predictor` that are only there for the interpret code, so we can't pull everything out...  That's a decision for later, though.

Reply:
https://github.com/allenai/allennlp-reading-comprehension/pull/11/commits/133c74160034779025438f319c46642c6f045053

Reply:
This kind of thing makes me worry about the whole project of splitting everything into separate repos. It's one of the pathologies of software development teams. Teams don't like the problems with their monorepo, and so they want to split it all up. But splitting it all up creates a mess of dependencies that wants to be managed.
- We have no way of running tests across repos. If I change something in the root `allennlp` library, I want the automated tests for my dependencies to run. Google has a system for this. We don't. We already struggled with this in Scala for S2, and never solved the problem.
- We can't guarantee that one version of `allennlp-reading-comprehension` always runs with a corresponding version of `allennlp`. So now we have to write backwards-compatible code every step of the way, or think hard about version pins for our own code. In practice, many projects will require work in multiple repos, so I now have a compatibility problem within a single project, and I have to carefully manage what gets merged when. This is why you had to have `allennlp-reading-comprehension` depend on a SHA. This is why the tests in that repo don't run right now.
- Circular dependencies, like we have in this case, always mean you want to split things up even more, making the problem worse. When you work on Hotflip2, do you want to coordinate changes in `allennlp`, `allennlp-reading-comprehension`, and `allennlp-interpret`?
- Back to tests: With lots of little repos, the ideal test is isolated to running in just one of them. It is hard to make a cross-repo test. So where do we stick integration tests? In my experience, unit tests are 90% useless, but integration tests are the ones that find bugs.

Reply:
1. We do have a way of running tests across repos: we run `allennlp-hub` tests on master commits, which @brendan-ai2 set up.  We run unit tests on all commits in a PR, integration tests on merge, in case our unit tests didn't catch something.  And when that happens, we make a new unit test that would have caught the thing that failed during integration tests.
2. Yeah, the sub-repos might sometimes have to pin things.  I think you're seeing issues now, though, because we're making breaking changes on purpose and we are in the middle of making things work.  Once 1.0 is released, the library should be stable and we will not make breaking changes very often, if at all, and this will be much less of an issue.
3. Yeah, the interpret stuff is more coupled with tasks than I would like it to be, so that particular one is a bit of an issue (and again is an argument for splitting it out, but that could easily make things more circular).  I don't think we will have this problem much in practice, though.  In general, splitting out the repos is also an attempt to force ourselves to think like users, where we have to use allennlp as a third-party library.  This should lead to better code that's easier for users, as we'll try to solve pain points when we run into them.
4. See answer to 1.